### PR TITLE
Add TransIP private network resource & attachment

### PR DESCRIPTION
**Pull request by JoooostB**
This resource makes use of a main resource and an attachment like other providers do as well. First you create a `private_network` resource after which you'd want to attach each VPS using a `private_network_attachment`

For example:

```
resource "transip_vps" "test" {
  name = "test"
  product_name = "vps-bladevps-x1"
  operating_system = "Debian 9"
}

resource "transip_private_network" {
    description = "test"
}

resource "transip_private_network_attachment" {
  private_network_id = transip_private_network.test.id
  vps_id = transip_vps.test.id

}
```

Please let me know if you notice any issues or bad practices, as this is my first experience with Go and/or extending Terraform Providers

### CODE REVIEW

File: data_source_transip_private_network.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+			},
+		},
+	}
+}
+
+func dataSourcePrivateNetworkRead(d *schema.ResourceData, m interface{}) error {
+	client := m.(repository.Client)
+	repository := vps.PrivateNetworkRepository{Client: client}
+	name := d.Get("name").(string)
+	// getPrivateNetworkID(d, m)
```
Comment (aequitas) [inline data_source_transip_private_network.go:48]: Commented out code can be removed if it's not used.

File: data_source_transip_private_network.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+	p, err := repository.GetByName(name)
+	if err != nil {
+		return fmt.Errorf("failed to lookup private network %q: %s", name, err)
+	}
+
+	log.Printf("[DEBUG] terraform-provider-transip success: private network %s found, getting details.\n", name)
```
Comment (aequitas) [inline data_source_transip_private_network.go:54]: I would reserve logging for stuff that still requires debugging sometimes, like complex logical statements or unreliable API responses. This seems straightforward code so all information you get from the debug log you already get when running the code as well, making the debug statement superfluous.

File: data_source_transip_private_network.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+	d.Set("description", p.Description)
+	log.Printf("[DEBUG] terraform-provider-transip success: private network %s has description %s \n", name, p.Description)
+	d.Set("is_blocked", p.IsBlocked)
+	d.Set("is_locked", p.IsLocked)
+	d.Set("vps_names", vpsNames)
+
+	return nil
+}
+
+// func getPrivateNetworkID(d *schema.ResourceData, m interface{}) error {
```
Comment (aequitas) [inline data_source_transip_private_network.go]: Same as above

File: data_source_transip_private_network_test.go [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/resource"
+)
+
+func TestAccTransipDataSourcePrivateNetwork(t *testing.T) {
+	var testConfig = `data "transip_private_network" "test" {name = "%s"}`
```
Comment (aequitas) [inline data_source_transip_private_network_test.go:12]: Please add a skip condition for when `TF_VAR_private_network_name` is not provided, like: https://github.com/aequitas/terraform-provider-transip/pull/16/files#diff-9b7b3d4debe01c7dce8fa350cddaed83R12-R15.

File: examples/private_network.tf [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+}
+
+# the private network data resource can be created by using the unique name TransIP assigns to the resource. (You can view this in the control panel)
+data "transip_private_network" "test" {
+    name = var.private_network_name
+}
+
+# order a new private network, or use `terraform import transip_private_netwerk.test test` to import existing one
+resource "transip_private_network" "test" {
+  description = "test" # Description is the user defined name, as the name attribute is set by TransIP as a unique ID.
```
Comment (aequitas) [inline examples/private_network.tf:12]: Good commenting on this difference between the data source and the resource.

File: provider_test.go [hunk 6]
```diff
@@ -39,4 +39,8 @@ func testAccPreCheck(t *testing.T) {
 	if v := os.Getenv("TF_VAR_domain"); v == "" {
 		t.Fatal("TF_VAR_domain must be set for acceptance tests")
 	}
+
+	if v := os.Getenv("TF_VAR_private_network_name"); v == "" {
+		t.Fatal("TF_VAR_private_network_name must be set for acceptance tests")
```
Comment (aequitas) [inline provider_test.go:44]: I would rather that we skip tests that require this variable, instead of disabling all tests if the variable is not provided. As not everyone (like me) will have a private netwerk configured. See earlier comment on how to enable per test skips.

File: resource_transip_private_network.go [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+	description := d.Get("description").(string)
+	client := m.(repository.Client)
+	repository := vps.PrivateNetworkRepository{Client: client}
+
+	err := repository.Order(description)
+	if err != nil {
+		return fmt.Errorf("failed to order private network %s: %s", description, err)
+
+	}
+	setPrivateNetworkID(d, m)
```
Comment (aequitas) [inline resource_transip_private_network.go:62]: Can you inline this code for now. I'm still trying to figure out what parts can be put into reusable functions. But for now I'd like all code in the function. 

Also you are ignoring the return value of `setPrivateNetworkID` which will break the functionality of the retry in case of a non-happy flow.

File: resource_transip_private_network.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+	"github.com/transip/gotransip/v6"
+	"github.com/transip/gotransip/v6/repository"
+	"github.com/transip/gotransip/v6/vps"
+)
+
+func resourcePrivateNetwork() *schema.Resource {
+	return &schema.Resource{
+		Create: resourcePrivateNetworkCreate,
+		Read:   resourcePrivateNetworkRead,
+		// Update: resourcePrivateNetworkUpdate,
```
Comment (aequitas) [inline resource_transip_private_network.go:18]: Does updating a private network make sense? As in, changing an attribute like description, without having to recreate the resource? In that case we could implement the update.

File: data_source_transip_private_network.go (hunk 1 above)
Comment (JoooostB) [reply data_source_transip_private_network.go]: I agree, removed above comment and the unused function it refers to.

File: resource_transip_private_network_attachment.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		log.Printf("[DEBUG] terraform-provider-transip trying to attach network %s to vps %s \n", privateNetworkID, vpsID)
+
+		err := repository.AttachVps(vpsID, privateNetworkID)
+		if err != nil {
+			if strings.Contains(err.Error(), fmt.Sprintf("VPS '%s' has an action running, no modification is allowed", vpsID)) {
+				log.Printf("[DEBUG] terraform-provider-transip VPS %s busy, retrying to attach to %s in a bit \n", vpsID, privateNetworkID)
+				return resource.RetryableError(fmt.Errorf("failed to attach VPS %s to private network %s, VPS busy: %s", vpsID, privateNetworkID, err))
+			}
+			log.Printf("[DEBUG] terraform-provider-transip something went wrong while attaching VPS %s to %s \n", vpsID, privateNetworkID)
```
Comment (aequitas) [inline resource_transip_private_network_attachment.go:58]: This debug message conveys the same information as the error message on the next line will. So it can be removed.

File: resource_transip_private_network_attachment.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+	client := m.(repository.Client)
+	repository := vps.PrivateNetworkRepository{Client: client}
+
+	return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		log.Printf("[DEBUG] terraform-provider-transip trying to attach network %s to vps %s \n", privateNetworkID, vpsID)
+
+		err := repository.AttachVps(vpsID, privateNetworkID)
+		if err != nil {
+			if strings.Contains(err.Error(), fmt.Sprintf("VPS '%s' has an action running, no modification is allowed", vpsID)) {
```
Comment (aequitas) [inline resource_transip_private_network_attachment.go:54]: Is there an other way we can deduct this is a temporary error from the API side? Like the HTTP return code being different? Comparing on the error string can be unreliable as it's not an API specification what text is returned.

File: resource_transip_private_network_attachment.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+	}
+
+	found := false
+	for _, vpsName := range p.VpsNames {
+		if vpsName == vpsID {
+			log.Printf("[DEBUG] terraform-provider-transip success: vps %s is the same as %s in private network %s \n", vpsID, vpsName, privateNetworkID)
+			found = true
+			break
+		}
+		if !found {
```
Comment (aequitas) [inline resource_transip_private_network_attachment.go:83]: I would move this out and below the for statement

File: resource_transip_private_network_attachment.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+func resourcePrivateNetworkAttachmentDelete(d *schema.ResourceData, m interface{}) error {
+	privateNetworkID := d.Get("private_network_id").(string)
+	vpsID := d.Get("vps_id").(string)
+
+	client := m.(repository.Client)
+	repository := vps.PrivateNetworkRepository{Client: client}
+	return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+		err := repository.DetachVps(vpsID, privateNetworkID)
+		if err != nil {
+			if strings.Contains(err.Error(), fmt.Sprintf("VPS '%s' has an action running, no modification is allowed", vpsID)) {
```
Comment (aequitas) [inline resource_transip_private_network_attachment.go:101]: Same as comment above about return code checking.

File: resource_transip_private_network_attachment_test.go [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+
+	"os"
+	"testing"
+)
+
+func TestAccTransipResourcePrivateNetworkAttachment(t *testing.T) {
+	testConfig := fmt.Sprintf(`
```
Comment (aequitas) [inline resource_transip_private_network_attachment_test.go:14]: Please also add skip conditions here as well.

File: resource_transip_private_network_attachment_test.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config:        testConfig,
+				ResourceName:  "transip_private_network_attachment.test",
+				ImportState:   false,
+				ImportStateId: os.Getenv("TF_VAR_private_network_name"),
```
Comment (aequitas) [inline resource_transip_private_network_attachment_test.go:29]: When using an environment variable in multiple places, it might be better to put it in a variable. This reduces the risk that a typo occurs.

File: resource_transip_private_network_test.go [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+	"github.com/hashicorp/terraform/helper/resource"
+
+	"os"
+	"testing"
+
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccTransipResourcePrivateNetwork(t *testing.T) {
+	testConfig := fmt.Sprintf(`
```
Comment (aequitas) [inline resource_transip_private_network_test.go:15]: Same comments as for `resource_transip_private_network_attachment_test.go`

File: resource_transip_vps.go [hunk 16]
```diff
@@ -153,24 +156,24 @@ func resourceVpsCreate(d *schema.ResourceData, m interface{}) error {
 
 	}
 
-	d.SetId(name)
-
 	d.Set("install_text", InstallText)
+	setVpsID(d, m)
```
Comment (aequitas) [inline resource_transip_vps.go:160]: Please inline this code, also ignoring return value.

File: resource_transip_vps.go [hunk 17]
```diff
@@ -153,24 +156,24 @@ func resourceVpsCreate(d *schema.ResourceData, m interface{}) error {
...
 
 	return resourceVpsRead(d, m)
 }
 
 func resourceVpsRead(d *schema.ResourceData, m interface{}) error {
-	name := d.Id()
+	name := d.Get("name").(string)
```
Comment (aequitas) [inline resource_transip_vps.go:166]: Can you explain a little better why this change is needed?

File: provider_test.go (hunk 6 above)
Comment (JoooostB) [reply provider_test.go:44]: I've changed `t.Fatal` to `t.Skip` here as well, would that suffice?

File: resource_transip_private_network.go (hunk 8 above)
Comment (JoooostB) [reply resource_transip_private_network.go:18]: I agree, it should be implemented as you may change your mind on the description :)

File: resource_transip_private_network_attachment.go (hunk 10 above)
Comment (JoooostB) [reply resource_transip_private_network_attachment.go:54]: Valid point, it actually returns a 409. Unfortunately it returns a 409 for multiple causes:

1. The PrivateNetwork 'example-privatenetwork' is already locked to another action
2. VPS 'example-vps' has an action running, no modification is allowed
3. VPS 'example-vps' is customer locked, no modification is allowed
4. Actions on VPS 'example-vps' are temporary disabled

So it's still pretty hard to conclude if we should continue or not. What's your suggestion on this?

File: resource_transip_private_network_attachment.go (hunk 12 above)
Comment (JoooostB) [reply resource_transip_private_network_attachment.go:101]: Please see reply

File: resource_transip_private_network_attachment_test.go (hunk 14 above)
Comment (JoooostB) [reply resource_transip_private_network_attachment_test.go:29]: I agree, it also looks nicer as well. Have been a bit inconsistent on this, will fix.

File: resource_transip_vps.go (hunk 16 above)
Comment (JoooostB) [reply resource_transip_vps.go:160]: I've moved it to be inline, but I'm not too sure about the return value of `resourceVpsCreate` now, could you check if this is right?

File: resource_transip_vps.go (hunk 17 above)
Comment (JoooostB) [reply resource_transip_vps.go:166]: It had something to do with the consistency of the unique ID TransIP has for your VPS and the description you give it in the resource. But reading it now, I think the inconsistency was in my head ;). Reverted it.

File: resource_transip_private_network_attachment.go (hunk 10 above)
Comment (aequitas) [reply resource_transip_private_network_attachment.go:52]: In this case it not that bad I think as it would only result in a retryable error becoming a fatal error. On the next Terraform run the resource would be properly detected as created as well (if enough time has passed). So lets keep it in for now.
